### ISSUE TITLE
Receive an error "KRB_ERR_RESPONSE_TOO_BIG" at  await client.Authenticate(credential); 

### ISSUE THREAD

**Issue by daniilzaonegin** [SteveSyfuhs/Kerberos.NET#220]
Hello!

I am writing DelegatingHandler for HttpClient to get Kerberos ticket and put it to www-authenticate header. And I've got some problems.
When I request Kerberos ticket I get Exception 

```c#
               using (var client = new KerberosClient(Krb5Config.Default()))
                {
                    var keyTable = new KeyTable(
                        _webEnv.ContentRootFileProvider.GetFileInfo(
                            "<keytab file name>").CreateReadStream());

                    var credential = 
                        new KeytabCredential("<account name>", 
                                keyTable, "<domain name>");
                    client.Configuration.Realms["<domain name>"].Kdc.Add("<server name>:88");

                    client.Configuration.Defaults.DefaultRealm = "<domain name>";
                    client.Configuration.DomainRealm.Add("<domain name>", "<domain name>");
                    client.Configuration.Defaults.DnsLookupKdc = false;

                    await client.Authenticate(credential); //exception occurs here
                    KrbApReq ticket = await client.GetServiceTicket("<spn name>");
                    wwwAuthValueAuth 
                        = Negotiate + Convert.ToBase64String(ticket.EncodeGssApi().Span);
                }

                request.Headers.Add(HeaderNames.WWWAuthenticate, wwwAuthValueAuth);
```
Exception StackTrace:
```
System.AggregateException: One or more errors occurred. (Cannot locate SRV record for RAIFFEISEN.RU) (KRB_ERR_RESPONSE_TOO_BIG: Response too big for UDP; retry with TCP) (Cannot locate a KDC Proxy endpoint for RAIFFEISEN.RU)
 ---> Kerberos.NET.Transport.KerberosTransportException: Cannot locate SRV record for RAIFFEISEN.RU
   at Kerberos.NET.Transport.KerberosTransportBase.LocatePreferredKdc(String domain, String servicePrefix) in D:\a\1\s\Kerberos.NET\Client\Transport\KerberosTransportBase.cs:line 130
   at Kerberos.NET.Transport.TcpKerberosTransport.GetClient(String domain) in D:\a\1\s\Kerberos.NET\Client\Transport\TcpKerberosTransport.cs:line 77
   at Kerberos.NET.Transport.TcpKerberosTransport.SendMessage[T](String domain, ReadOnlyMemory`1 encoded, CancellationToken cancellation) in D:\a\1\s\Kerberos.NET\Client\Transport\TcpKerberosTransport.cs:line 53
   at Kerberos.NET.Transport.KerberosTransportSelector.SendMessage[T](String domain, ReadOnlyMemory`1 encoded, CancellationToken cancellation)
   --- End of inner exception stack trace ---
   at Kerberos.NET.Transport.KerberosTransportSelector.SendMessage[T](String domain, ReadOnlyMemory`1 encoded, CancellationToken cancellation) in D:\a\1\s\Kerberos.NET\Client\Transport\KerberosTransportSelector.cs:line 86
   at Kerberos.NET.Client.KerberosClient.RequestTgt(KerberosCredential credential) in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 1009
   at Kerberos.NET.Client.KerberosClient.AuthenticateCredential(KerberosCredential credential) in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 345
   at Kerberos.NET.Client.KerberosClient.Authenticate(KerberosCredential credential) in D:\a\1\s\Kerberos.NET\Client\KerberosClient.cs:line 328
```

Could you explain why this happens and how I can resolve it?

Thank you!

**Comment by daniilzaonegin** [comment]
I found out that `KRB_ERR_RESPONSE_TOO_BIG` could be fixed by changing parameter ` client.Configuration.Defaults.UdpPreferenceLimit`. Errors `(Cannot locate SRV record for RAIFFEISEN.RU) (Cannot locate a KDC Proxy endpoint for RAIFFEISEN.RU)` remain.

**Comment by daniilzaonegin** [comment]
After investigation I realized, the problem is that SRV DNS record not found by name `_kerberos._udp/_kerberos._tcp` (https://web.mit.edu/kerberos/krb5-1.4/krb5-1.4.1/doc/krb5-admin/Hostnames-for-KDCs.html). Is there any way to specify this name explicitly without DNS lookup? Like in krb5.conf https://web.mit.edu/kerberos/krb5-1.12/doc/admin/conf_files/krb5_conf.html.

**Comment by SteveSyfuhs** [comment]
Your configuration appears correct. That should coerce it into not querying DNS at all. Let me see if I can repro it.

________________________________

**Comment by daniilzaonegin** [comment]
ok 👍  Thanks!

**Comment by SteveSyfuhs** [comment]
Okay, so I figured out the issue. You're hitting a funny (not haha-funny, but I-swear-this-worked-before-funny) bug where the `host:port` value is being treated as a URI in `ParseKdcEntryAsSrvRecord` in `ClientDomainService`.

https://github.com/dotnet/Kerberos.NET/blob/8c254ce2a847248ae4257595e3162212397e4f01/Kerberos.NET/Client/Transport/ClientDomainService.cs#L177-L185

Tl;dr; drop the `:88` from the configuration. This library will use port 88 by default.

Note, if you do have to include a non-standard port you can maybe work around this using `client.PinKdc(...)`

So what's happening is

1. Try TCP
   1. Get list of configured hosts
   2. Parse strings
   3. Parse as URI -- whoops
   4. Find no TCP-compatible hosts, throw "Cannot locate SRV record for..."
2. Try UDP
   1. Is UDP even configured? No, throw KRB_ERR_RESPONSE_TOO_BIG
3. Try HTTP
   1. Get list of configured hosts
   2. Parse strings
   3. Does parsed string actually start with HTTP or HTTPS? -- ignore
   4. Find no HTTP-compatible hosts, throw "Cannot locate a KDC Proxy endpoint for..."
  
And so it raises each of the three thrown errors.

**Comment by SteveSyfuhs** [comment]
Also, the way you've written the code will in fact guarantee it never tries DNS, it just happens that the errors raised are identical if it can't find the SRV record by DNS or by config.

**Comment by daniilzaonegin** [comment]
:) thank you! After dropping port, I got kerberos ticket! But now my api server returns http-401. I look into that tomorrow, but this is another issue. It could be, that I need keytab file for another SPN.
